lib/semantics/tools.cc

```diff
@@ -443,6 +490,45 @@ bool IsAssumedSizeArray(const Symbol &symbol) {
   return details && details->IsAssumedSize();
 }
 
+bool IsExternalInPureContext(const Symbol &symbol, const Scope &scope) {
+  if (const auto *pureProc{semantics::FindPureProcedureContaining(&scope)}) {
+    if (const Symbol * root{GetAssociationRoot(symbol)}) {
+      if (semantics::FindExternallyVisibleObject(*root, *pureProc)) {
+        return true;
+      }
+    }
+  }
+  return false;
+}
+
+bool InProtectedContext(const Symbol &symbol, const Scope &currentScope) {
+  return IsProtected(symbol) && !IsHostAssociated(symbol, currentScope);
+}
+
+std::optional<parser::MessageFixedText> WhyNotModifiable(
+    const Symbol &symbol, const Scope &scope) {
+  const Symbol *root{GetAssociationRoot(symbol)};
+  if (!root) {
+    return "'%s' is construct associated with an expression"_en_US;
+  } else if (InProtectedContext(*root, scope)) {
+    return "'%s' is protected in this scope"_en_US;
+  } else if (IsExternalInPureContext(*root, scope)) {
+    return "'%s' is an external that is referenced in a PURE procedure"_en_US;
```

Comment:
"External" usually refers to an external procedure so maybe this would be better:
"'%s' is externally visible and referenced in a PURE procedure"

Reply:
C1594 has some other restrictions on what can be modified in a pure procedure, e.g. something in a common block or a co-indexed object.

Reply:
I like your rewording of the message.

With regards to C1594, it looks like I just spaced on this.  I have a list on Google docs of all of the places where various constructs are prohibited from appearing in variable definition contexts, and this document already lists C1594.  It also lists C1101 and C1158 neither of which I currently check.  I have more work to do ...

Reply:
More on C1594 ...  It looks to me like all of the cases listed in the constraint are handled by ```FindExternallyVisibleObject()``` except for a coindexed object.  I added a TODO for that case.